src/promql/src/planner.rs

```diff
@@ -0,0 +1,627 @@
+    use query::query_engine::QueryEngineState;
+    use query::DfContextProviderAdapter;
+    use session::context::QueryContext;
+    use table::metadata::{TableInfoBuilder, TableMetaBuilder};
+    use table::test_util::EmptyTable;
+
+    use super::*;
+
+    async fn build_test_context_provider(
+        table_name: String,
+        num_tag: usize,
+        num_field: usize,
+    ) -> DfContextProviderAdapter {
+        let mut columns = vec![];
+        for i in 0..num_tag {
+            columns.push(ColumnSchema::new(
+                format!("tag_{i}"),
+                ConcreteDataType::string_datatype(),
+                false,
+            ));
+        }
+        columns.push(
+            ColumnSchema::new(
+                "timestamp".to_string(),
+                ConcreteDataType::timestamp_millisecond_datatype(),
+                false,
+            )
+            .with_time_index(true),
+        );
+        for i in 0..num_field {
+            columns.push(ColumnSchema::new(
+                format!("field_{i}"),
+                ConcreteDataType::float64_datatype(),
+                true,
+            ));
+        }
+        let schema = Arc::new(Schema::new(columns));
+        let table_meta = TableMetaBuilder::default()
+            .schema(schema)
+            .primary_key_indices((0..num_tag).collect())
+            .value_indices((num_tag + 1..num_tag + 1 + num_field).collect())
+            .next_column_id(1024)
+            .build()
+            .unwrap();
+        let table_info = TableInfoBuilder::default()
+            .name(&table_name)
+            .meta(table_meta)
+            .build()
+            .unwrap();
+        let table = Arc::new(EmptyTable::from_table_info(&table_info));
+        let catalog_list = Arc::new(MemoryCatalogManager::default());
+        catalog_list
+            .register_table(RegisterTableRequest {
+                catalog: DEFAULT_CATALOG_NAME.to_string(),
+                schema: DEFAULT_SCHEMA_NAME.to_string(),
+                table_name,
+                table_id: 1024,
+                table,
+            })
+            .await
+            .unwrap();
+
+        let query_engine_state = QueryEngineState::new(catalog_list);
+        let query_context = QueryContext::new();
+        DfContextProviderAdapter::new(query_engine_state, query_context.into())
+    }
+
+    // {
+    // 	input: `abs(some_metric{foo!="bar"})`,
+    // 	expected: &Call{
+    // 		Func: MustGetFunction("abs"),
+    // 		Args: Expressions{
+    // 			&VectorSelector{
+    // 				Name: "some_metric",
+    // 				LabelMatchers: []*labels.Matcher{
+    // 					MustLabelMatcher(labels.MatchNotEqual, "foo", "bar"),
+    // 					MustLabelMatcher(labels.MatchEqual, model.MetricNameLabel, "some_metric"),
+    // 				},
+    // 			},
+    // 		},
+    // 	},
+    // },
+    async fn do_single_instant_function_call(fn_name: &'static str, plan_name: &str) {
+        let prom_expr = PromExpr::Call {
```

Comment:
Why not just use promql-parser to parse the expression and get this expr?

Reply:
well... I think it's not ready yet. Currently only lexer is available IIRC